src/qibolab/instruments/zhinst/sweep.py

```diff
@@ -0,0 +1,139 @@
+            for pulse in sweeper.pulses or []:
+                if sweeper.parameter in (Parameter.duration, Parameter.start):
+                    sweep_param = lo.SweepParameter(
+                        values=sweeper.values * NANO_TO_SECONDS
+                    )
+                    pulse_sweeps.append((pulse, sweeper.parameter, sweep_param))
+                elif sweeper.parameter is Parameter.frequency:
+                    ptype, qubit = pulse.type, qubits[pulse.qubit]
+                    if ptype is PulseType.READOUT:
+                        ch = measure_channel_name(qubit)
+                        intermediate_frequency = (
+                            qubit.readout_frequency
+                            - qubit.readout.local_oscillator.frequency
+                        )
+                    elif ptype is PulseType.DRIVE:
+                        ch = qubit.drive.name
+                        intermediate_frequency = (
+                            qubit.drive_frequency
+                            - qubit.drive.local_oscillator.frequency
+                        )
+                    else:
+                        raise ValueError(
+                            f"Cannot sweep frequency of pulse of type {ptype}, because it does not have associated frequency"
+                        )
+                    sweep_param = lo.SweepParameter(
+                        values=sweeper.values + intermediate_frequency
+                    )
+                    channel_sweeps.append((ch, sweeper.parameter, sweep_param))
+                elif (
+                    pulse.type is PulseType.READOUT
+                    and sweeper.parameter is Parameter.amplitude
+                ):
+                    sweep_param = lo.SweepParameter(
+                        values=sweeper.values / max(sweeper.values)
+                    )
+                    channel_sweeps.append(
+                        (
+                            measure_channel_name(qubits[pulse.qubit]),
+                            sweeper.parameter,
+                            sweep_param,
+                        )
+                    )
+                else:
+                    sweep_param = lo.SweepParameter(values=copy(sweeper.values))
+                    pulse_sweeps.append((pulse, sweeper.parameter, sweep_param))
+                parallel_sweeps.append((sweeper, sweep_param))
+
+            for qubit in sweeper.qubits or []:
+                if sweeper.parameter is not Parameter.bias:
+                    raise ValueError(
+                        f"Sweeping {sweeper.parameter.name} for {qubit} is not supported"
+                    )
+                sweep_param = lo.SweepParameter(
+                    values=sweeper.values + qubit.flux.offset
+                )
+                channel_sweeps.append((qubit.flux.name, sweeper.parameter, sweep_param))
+                parallel_sweeps.append((sweeper, sweep_param))
+
+            for coupler in sweeper.couplers or []:
```

Comment:
Do you really need to treat `sweeper.couplers` separate from `sweeper.qubits`?

I believe the operations to be identical, so you could just concatenate the two iterators.
(this would help the effort of handling couplers together with qubits, as much as possible)